content/200-concepts/100-components/02-prisma-client/053-client-extensions/201-custom-validation.mdx

```diff
@@ -0,0 +1,133 @@
+  // first name is between 2 and 50 characters long
+  firstName: z.string().min(2).max(50),
+  // first name is between 2 and 50 characters long
+  lastName: z.string().min(2).max(50),
+}) satisfies z.Schema<Prisma.UserCreateInput>
+
+const prisma = new PrismaClient()
+  .$extends({
+    /**
+     * The `query` component allows you to hook into a query and modify query arguments
+     */
+    query: {
+      user: {
+        create({ args, query }) {
+          /** validate the user input based on the schema */
+          args.data = UserSchema.parse(args.data);
+          return query(args)
+        },
+        update({ args, query }) {
+          /** the partial() method marks all the properties in the schema as optional */
+          args.data = UserSchema.partial().parse(args.data);
+          return query(args)
+        },
+      }
+    }
+  })
+
+// Example query
+await prisma.user.create({
+  data: {
+    firstName: 'Alice',
+    lastName: 'Jackson',
+    password: 'randomnumbers',
+    email: 'alice@prisma.io',
+  }
+})
+
+// This query will fail because it the email input is invalid
+await prisma.user.update({
+  where: {
+    id: 1
+  },
+  data: {
+    email: 'user@prisma'
+  }
+})
+```
+
+The Node.js ecosystem offers a number of high-quality, easy-to-use validation libraries to choose from including: [joi](https://github.com/sideway/joi), [validator.js](https://github.com/validatorjs/validator.js), [Yup](https://github.com/jquense/yup), [Zod](https://github.com/colinhacks/zod) and [Superstruct](https://github.com/ianstormtaylor/superstruct).
+
+The example is based on the following sample schema:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["clientExtensions"]
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+model Post {
+  id        Int     @id @default(autoincrement())
+  title     String
+  content   String?
+  published Boolean @default(false)
+
+  user     User? @relation(fields: [authorId], references: [id])
+  authorId Int?
+}
+
+model User {
+  id    Int     @id @default(autoincrement())
+  name  String?
+  email String  @unique
+
+  posts Post[]
+}
+```
+
+</TopBlock>
+
+## Going further
+
+- Learn more about [Prisma Client extensions](/concepts/components/prisma-client/client-extensions).
+- Learn more sample use cases you could use Prisma Client extensions for in [this blog post](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn).
+- You can find more Prisma Client extensions sample cases in this [GitHub repository](https://github.com/prisma/prisma-client-extensions)
+
+{/\*\*
+
+## Going Further
+
+- Learn how you can create and modify queries using [Prisma Client extensions](/concepts/components/prisma-client/client-extensions/query)
+- Learn how you can organize your code better by moving the `signup` function into [a custom model](/concepts/components/prisma-client/custom-models).
+  \*/}
```

Comment:
Broken output:
<img width="584" alt="image" src="https://user-images.githubusercontent.com/183673/221590063-140d6655-3e7f-4eb4-bb46-dd3642bed476.png">

Reply:
Came here to note the same 🙃